### PR TITLE
feat(common/models): 'revert' now uses model's punctuation

### PR DESCRIPTION
**Pull request by jahorton**
Feature:
- The 'reversion' option for undoing applied suggestions now uses the current lexical model's punctuation settings.

Other notes:
- Adds unit tests for the generated reversion
- Adds unit tests for `Transform` merging, given the new `buildMergedTransform` signature and implementation.

To demonstrate the principle feature, I made a temporary and slightly edited version of the MTNT model, as seen here:

Screen Shot 2020-10-02 at 9 49 43 AM

Note the "<<" and ">>" syntax, instead of the MTNT model's original use of double-quotes.  Previously, 'revert' options always used double-quotes, so this provides the visual distinction necessary to see the difference at a glance.

Selecting the highlighted suggestion, as usual, displays new suggestions:

Screen Shot 2020-10-02 at 9 51 20 AM

Hitting BKSP then triggers the option to revert:

Screen Shot 2020-10-02 at 9 54 21 AM

### CODE REVIEW

File: common/models/types/index.d.ts [hunk 1]
```diff
@@ -238,6 +238,8 @@ declare interface Suggestion {
   tag?: SuggestionTag;
 }
 
+interface Reversion extends Suggestion { }
```
Comment (eddieantonio) [inline common/models/types/index.d.ts:241]: Is this supposed to be the following?

```suggestion
interface Reversion extends Suggestion {
  tag: 'revert';
}
```

File: common/predictive-text/docs/worker-communication-protocol.md [hunk 2]
```diff
@@ -459,6 +463,40 @@ point.  If no such wordform exists, this returns an empty string.
...
+necessary to revert acceptance of the `suggestion`.
+
+### Message: `postaccept`
+
+The `postaccept` message is sent from the keyboard to the LMLayer.  This
+message sends a `reversion` capable of undoing acceptance of the `suggestion`
+just accepted by the triggering `accept` message.
+
+This message **MUST** be in response to a `postaccept` message, and it
+**MUST** respond with the corresponding [token][Tokens].
```
Comment (eddieantonio) [inline common/predictive-text/docs/worker-communication-protocol.md:498]: 👍

File: common/predictive-text/docs/worker-communication-protocol.md [hunk 3]
```diff
@@ -459,6 +463,40 @@ point.  If no such wordform exists, this returns an empty string.
...
+
+The `accept` message is sent from the keyboard to the LMLayer.  This
+message tells the LMLayer that a previously-returned `suggestion` 
+(from a `predict`-`suggestions` pair) has been accepted by the user.  
+The LMLayer **SHOULD** respond to each `accept` message with a 
+`postaccept` message providing a `reversion` capable of undoing it.
+
+The keyboard **MUST** send the `suggestion` and `context` parameters.
+The keyboard **MUST** send a unique token.  The `postTransform` parameter is 
+optional, but highly suggested.
```
Comment (eddieantonio) [inline common/predictive-text/docs/worker-communication-protocol.md:476]: Can you please elaborate on `postTransform` in the documentation?

File: common/predictive-text/worker/worker-interfaces.ts [hunk 4]
```diff
@@ -118,6 +118,37 @@ interface WordbreakMessage {
...
+   */
+  context: Context;
+
+  /**
+   * A Transform representing any text manipulations applied to 
+   * the Context after the `suggestion` was generated.  
+   * 
+   * Necessary, as Suggestions are generated without applying their
+   * triggering keystroke to the Context.  (The current context is 
+   * thus likely to differ.)
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts:147]: Makes sense!

**Review by eddieantonio — APPROVED** [review]
LGTM! I'd just like a short sentence about `postTransform` in the worker communication protocol docs.

File: common/models/types/index.d.ts (hunk 1 above)
Comment (jahorton) [reply common/models/types/index.d.ts:241]: I mean, I didn't realize it at the time... but yes, it _is_ supposed to be that.  Good idea.

File: common/predictive-text/docs/worker-communication-protocol.md (hunk 3 above)
Comment (jahorton) [reply common/predictive-text/docs/worker-communication-protocol.md:476]: Done.  I added it within the section beneath this, as it was easier to keep brief there.

**Pull request merged**